### PR TITLE
filter.go: add generic filtering mechanism

### PR DESCRIPTION
**Pull request by foolusion**
This adds a Select function that takes a slice of components and a
ComponentPredicate function. This allows both AND and OR comparisons.

### CODE REVIEW

**Comment by Kashomon** [comment]
Needs some tests, but I like the idea. Is the plan to eventually remove the old filtering logic?

File: pkg/filter/filter.go [hunk 1]
```diff
@@ -297,3 +297,178 @@ func matches(d *objectData, o *Options) bool {
...
+				}
+			}
+		}
+		return false
+	}
+}
+
+// UnstructuredNamespaceIn checks if any of the components object.GetNamespace()
+// are in namespaces.
+func UnstructuredNamespaceIn(namespaces ...string) ComponentPredicate {
```
Comment (Kashomon) [inline pkg/filter/filter.go:334]: Maybe "ObjectNamespaceIn" in components, the unstructured resources are generally referred to as Objects.

**Comment by foolusion** [comment]
> Needs some tests, but I like the idea. Is the plan to eventually remove the old filtering logic?

I don't know, we could convert the internal logic to use this, but replacing it would break API compatibility. I think we also need to write a similar SelectObjects method that returns a list of objects.

File: pkg/filter/filter.go (hunk 1 above)
Comment (foolusion) [reply pkg/filter/filter.go]: Done

File: pkg/filter/filter.go [hunk 2]
```diff
@@ -297,3 +297,202 @@ func matches(d *objectData, o *Options) bool {
...
+				return true
+			}
+		}
+		return false
+	}
+}
+
+// And returns true if every ComponentPredicate function returns true and
+// returns false otherwise.
+func And(componentPredicates ...ComponentPredicate) ComponentPredicate {
```
Comment (Kashomon) [inline pkg/filter/filter.go:453]: Do you need a Not method? Otherwise maybe add a Filter method?

File: pkg/filter/example_select_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+func toComponentList(components []*bundle.Component) []component {
+	out := make([]component, len(components))
+	for i, c := range components {
+		out[i] = toComponent(c)
+	}
+	return out
+}
+
+func Example_select() {
```
Comment (Kashomon) [inline pkg/filter/example_select_test.go:116]: This isn't a test method...?

Also, I would recommend a table test that has more extensive cases.

File: pkg/filter/BUILD.bazel [hunk 4]
```diff
@@ -15,7 +15,11 @@ go_library(
 
 go_test(
     name = "go_default_test",
-    srcs = ["filter_test.go"],
+    srcs = [
+        "example_selectObjects_test.go",
```
Comment (Kashomon) [inline pkg/filter/BUILD.bazel:19]: not seeing this file.

File: pkg/filter/example_selectObjects_test.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package filter_test
```
Comment (Kashomon) [inline pkg/filter/example_selectObjects_test.go:1]: file comment: please make all files snake_case.go

File: pkg/filter/example_select_test.go (hunk 3 above)
Comment (Kashomon) [reply pkg/filter/example_select_test.go:116]: Oh, I see, it's an example. I've honestly never used examples.

Still, some table tests might be nice.

**Comment by Kashomon** [comment]
/lgtm

I'll leave it to you to merge if you feel you want to do more polishing.

**Review by Kashomon — APPROVED** [review]

**Pull request merged**